plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php

```diff
@@ -0,0 +1,217 @@
+	 */
+	public function __construct( array $initial_resolved_cache ) {
+		$this->resolved_cache = $initial_resolved_cache;
+	}
+
+	/**
+	 * Get an instance of a class.
+	 *
+	 * ContainerException will be thrown in these cases:
+	 *
+	 * - $class_name is outside the WooCommerce root namespace (and wasn't included in the initial resolve cache).
+	 * - The class referred by $class_name doesn't exist.
+	 * - Recursive resolution condition found.
+	 * - Reflection exception thrown when instantiating or initializing the class.
+	 *
+	 * A "recursive resolution condition" happens when class A depends on class B and at the same time class B depends on class A, directly or indirectly;
+	 * without proper handling this would lead to an infinite loop.
+	 *
+	 * Note that this method throwing ContainerException implies that code fixes are needed, it's not an error condition that's recoverable at runtime.
+	 *
+	 * @param string $class_name The class name.
+	 *
+	 * @return object The instance of the requested class.
+	 * @throws ContainerException Error when resolving the class to an object instance.
+	 * @throws \Exception Exception thrown in the constructor or in the 'init' method of one of the resolved classes.
+	 */
+	public function get( string $class_name ) {
+		$class_name    = trim( $class_name, '\\' );
+		$resolve_chain = array();
+		return $this->get_core( $class_name, $resolve_chain );
+	}
+
+	/**
+	 * Core function to get an instance of a class.
+	 *
+	 * @param string $class_name The class name.
+	 * @param array  $resolve_chain Classes already resolved in this resolution chain. Passed between recursive calls to the method in order to detect a recursive resolution condition.
+	 * @return object The resolved object.
+	 * @throws ContainerException Error when resolving the class to an object instance.
+	 */
+	protected function get_core( string $class_name, array &$resolve_chain ) {
+		if ( isset( $this->resolved_cache[ $class_name ] ) ) {
+			return $this->resolved_cache[ $class_name ];
+		}
+
+		// phpcs:disable WordPress.Security.EscapeOutput.ExceptionNotEscaped
+
+		if ( ! $this->is_class_allowed( $class_name ) ) {
+			throw new ContainerException( "Attempt to get an instance of class '$class_name', which is not in the " . self::WOOCOMMERCE_NAMESPACE . ' namespace. Did you forget to add a namespace import?' );
+		}
+
+		if ( ! class_exists( $class_name ) ) {
+			throw new ContainerException( "Attempt to get an instance of class '$class_name', which doesn't exist." );
+		}
+
+		// Account for the containers used by the Store API and Blocks.
+		if ( StringUtil::starts_with( $class_name, 'Automattic\WooCommerce\StoreApi\\' ) ) {
+			return StoreApi::container()->get( $class_name );
+		}
+		if ( StringUtil::starts_with( $class_name, 'Automattic\WooCommerce\Blocks\\' ) ) {
+			return BlocksPackage::container()->get( $class_name );
+		}
+
+		if ( in_array( $class_name, $resolve_chain, true ) ) {
```

Comment:
Do you have an idea of how big the `$resolve_chain` may get?  If it frequently gets a couple dozen items, it may make sense, for performance, to use a keyed array instead, e.g. `$resolved_set[ $class_name ] = true`. 

- Once the list gets so large, the cost of creating the hash is cheaper than the iterations.
- It doesn't look like the order is needed during the recursion, so a stack isn't necessary.
- All of the classes are going to start with the same prefixes so more of every class name will have to be checked.

Can we at least make `::get_core()` private so we can pivot in the future without the worry of backward compatibility.  Or, even better, make it a final class until we have a reason for it not to be.

Reply:
> Do you have an idea of how big the $resolve_chain may get?

There are a couple of convoluted cases (related to HPOS) in which the array gets 16 items, but otherwise it never gets more than 5.

> Can we at least make ::get_core() private

I'll probably need `get_core` to be accessible from the `TestingContainer` class I'll develop. But don't worry about this: the class is in the `Internal` namespace so we don't have to keep backwards compatibility.

Reply:
This has happened too with [the batch processing engine](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php), but in both cases the "external" code is still managed by us (Automattic/WooCommerce). And while we must be careful with any changes indeed, the point is that to the external world (extension developers) this code is private for all intents and purposes, it's documented as such, and we have freedom to change it whenever we need to do so.

Reply:
Hi @Konamiman, great initiative first of all! I'm reviewing the changes, but before that I just wanted to say I'm a bit with @prettyboymp here, what third-party shouldn't do but can technically do has not proven to be a reliable binding contract :/
Especially with access modifiers, not getting them locked in to start with might bite us in the future, and prevent changes in the future without breaking stores. Ultimately it is third-party responsibility to keep up with us, but that usually ends with a code freeze exception on our end :(

Reply:
We'll do a public announcement about this, including the mehcanism to switch to the old container and its removal in WooCommerce 10.0, right after this is merged. This should be enough, we've done a similar thing with the removal of the legacy REST API (which _was_ an actual breaking change) and it went well.

As for our canonical extensions, see [this comment](https://github.com/woocommerce/woocommerce/pull/52296#discussion_r1836093576).